Vectorize normal_

goldsborough opened the pull request:
PyTorch is known to have high start-up times, to a large part due to slow tensor initialization/filling when calling `normal_`. [Torch's `normal` function](https://github.com/pytorch/pytorch/blob/master/aten/src/TH/THRandom.c#L253) uses the Box-Mueller transform to produce Gaussian distributed floats from uniform floats. I did some benchmarks and found that generating normal numbers took around 5 times longer than generating only uniform floats, suggesting that the current normal sampling code was the bottleneck.

This PR addresses this by introducing a vectorized version of the Box-Mueller transform that essentially does the same thing, but for 8 values at a time. This version is called only for floats, only if we have AVX2, only if there are more than 16 values (due to implementation) and only if the tensor is contiguous. However, this should cover like 90%-95% of real-world cases where it's currently slow.

My initial, small-scale benchmarks show a 5x-6x speed-up:

Before:
```
In [1]: import torch
In [2]: x = torch.Tensor(10000, 10000)
In [3]: %time _ = x.normal_(0, 1)
CPU times: user 3.45 s, sys: 111 ms, total: 3.57 s
Wall time: 3.57 s
```

After:
```
In [1]: import torch
In [2]: x = torch.Tensor(10000, 10000)
In [3]: %time _ = x.normal_(0, 1)
CPU times: user 611 ms, sys: 1.07 ms, total: 612 ms
Wall time: 613 ms
```

Which looks pretty good. I will see how this affects loading a model like imagenet or similar.

CC @zdevito

goldsborough committed:
Vectorize normal_

apaszke reviewed (commented):
This looks good, I only have two minor comments.

My only concern is that it doesn't only vectorize the function, but it will also make it return different results on AVX2 and non-AVX2 platforms for the same random seed. Not sure how much do we care about cross-platform reproducibility, since that's very hard and constraining, but it's worth noting that. `normal` is quite important because it's used for initializing weights. @soumith thoughts?

apaszke commented:
Oh, also it would be nice to evaluate some of the real examples (e.g. word language model form our repo), just to make sure we won't get severe downclocking penalty because of AVX2.

goldsborough commented:
Thanks for the comments. About to board a flight, so will fix comments in a bit. Just a note: the non-AVX2 code can be re-written to layout the numbers in the same order as the AVX2 version (8 at a time, interleaved). Something like https://gist.github.com/goldsborough/75ee1802110eda71517cc33ea3c59a88

zdevito reviewed (approved):
This looks great!

Changing the serial version to mimic the AVX one seems like a good idea since it increases the perf. of the serial one and also makes it the same as the avx one.

colesbury commented on aten/src/TH/generic/THTensorRandom.c:
```diff
@@ -92,6 +92,17 @@ void THTensor_(uniform)(THTensor *self, THGenerator *_generator, double a, doubl
+#if defined(TH_REAL_IS_FLOAT) && defined(USE_AVX2)
+  const int size = THTensor_(numel)(self);
+  if (size >= 16 && THTensor_(isContiguous)(self)) {
+    THTensor_(normal_fill_AVX2)(self->storage->data,
```
Does this handle alignment of `data`?

I think we need cpu dispatch of `normal_fill_AVX2`. For example, we distribute binary builds that are compiled with AVX, but need to run on processors without AVX2 support.

colesbury replied:
See THVectorDispatch.c/h for example

soumith replied:
^^ what Sam says. see vectorDispatchInit in https://github.com/pytorch/pytorch/blob/master/aten/src/TH/generic/THVectorDispatch.c

soumith reviewed (commented):
looks pretty good, needs some runtime dispatch changes (see comments in-line)

goldsborough committed:
Add interleaved scalar normal fill function and implement proper dispatch

goldsborough committed:
Use fmadd instead of multiply + add

goldsborough commented:
## Changes

Latest 2 commits make the following changes:

- Actually use the mean and standard deviation (was not using the variables before and thus generating unit Gaussian samples),
- Added a scalar `normal_fill` function that interleaves values just like the vectorized code, so that there is no difference in generated samples between AVX and non-AVX platforms. This function is also around 1.5x faster than the old version. Thus also non-AVX versions get a speedup (for contiguous tensors with at least 16 values).
- Implemented all the vector dispatch stuff
- Using `int64_t` for size
- Using `THAssert` instead of `assert`
- AVX code uses `_m256_loadu_ps` instead of `_m256_load_ps`, i.e. misaligned loads so that it works also for misaligned data,
- Using explicit FMA `_m256_fmadd_ps` instruction instead of multiply + add (clang doesn't do this automatically, GCC does ... it's less code anyway. Also FMA is available wherever AVX is on Intel and AMD chips, so should be fine). Marginally faster.

## Benchmarks

This time compiling with GCC 7.

### Microbenchmarks

10,000 x 10,000 (float/AVX): 3.3s -> 0.48s (6.875x speedup)
1,000 x 1,000 (float/AVX): 35ms -> 4.9ms (7.1x speedup)
10,000 x 10,000 (double/scalar): 3.2s -> 2.1s (1.5x speedup)

float/AVX here means it's using the vectorized version, double/scalar means it's the interleaved scalar function that I added, since the vectorized version is only called for floats.

### Imagenet Startup Times

VGG19: 5.61s -> 1.62s (3.5x speedup)
ResNet101: 2.4s -> 0.65s (3.7x speedup)
ResNet50: 1.21s -> 0.46s (2.6x speedup)

Happy new year 🎆 :tada:

soumith commented:
@pytorchbot add to whitelist

apaszke commented on aten/src/TH/vector/AVX2.c:
```diff
@@ -45,4 +47,62 @@ void THFloatVector_cadd_AVX2(float *z, const float *x, const float *y, const flo
+
+  if (size % 16 != 0) {
+    // We rewind so that we have 16 values and then compute them in one step.
+    data += (size - 16);
```
I think in here `data` still points to the beginning of the buffer (you've been doing `data + i`, not `data += 16` in the loop above). Shouldn't this be `data = data + size - 16`? It might be a tiny bit cleaner and safer to declare that as a new pointer instead of overwriting this one

goldsborough replied:
You're right that it's still pointing to the start. The idea here is to place it to the start of the last 16 values. So I think it's ok, but I'll apply your suggestion of `data = data + size - 16`.

apaszke replied:
Ohhh right. Not sure why I thought it was that way

apaszke reviewed (commented):
LGTM! I think there's one small bug though

goldsborough committed:
Be more explicit about pointer increment

apaszke commented:
Should be ready to merge but the builds are failing now (probably because the seed we used previously is unlucky after this change)

goldsborough commented:
Ok, how do we resolve the build failure? It says something about not being able to pull the changes.

apaszke commented:
Oh this one looks like a CI failure, but the CUDA builds manage to build and fail at test time

zdevito commented:
@pytorchbot retest this please

goldsborough commented:
No luck with random seeds for cudnn builds. I will need a gpu machine to figure out what's wrong locally. Or is there any smarter way of solving these random failures other than finding a lucky seed?

apaszke commented:
Alright, I looked into it and it seems that the test that's failing now is particularily flaky when using half (63 failures / 1000 trials). Reducing the scale of values used to test it makes it more precise, and it succeeded 10000 times now. Here's the patch (just add `/ 2` in two places):
```patch
--- a/test/test_nn.py                                                                    
+++ b/test/test_nn.py                                                                    
@@ -2132,9 +2132,9 @@ class TestNN(NNTestCase):                                          
                 continue                                                                
             for depth_multiplier in [1, 2]:                                             
                 m = nn.Conv2d(2, 2 * depth_multiplier, kernel_size=3, groups=2).type(tp)
-                i = Variable(torch.randn(2, 2, 6, 6).type(tp), requires_grad=True)      
+                i = Variable(torch.randn(2, 2, 6, 6).type(tp) / 2, requires_grad=True)  
                 output = m(i)                                                           
-                grad_output = torch.randn(2, 2 * depth_multiplier, 4, 4).type(tp)       
+                grad_output = torch.randn(2, 2 * depth_multiplier, 4, 4).type(tp) / 2   
                 output.backward(grad_output)                                            
                                                                                         
                 offset = 1 * depth_multiplier                                                                                      
```

goldsborough committed:
Stabilizy flaky test

goldsborough commented:
That worked, thanks Adam! Now one of the builds got stuck after a segfault in the dataloader tests. Seems flaky too.

goldsborough commented:
😍

apaszke commented:
There's already other PR open that fixes the data loader thing. The test had a race condition

apaszke commented:
Thanks Peter!

aluo-x commented:
Not sure if this warrents a new bug report. But today while trying to build Pytorch on Windows, I ran into the following error:

```
"C:\optimae\pytorch\torch\lib\build\ATen\INSTALL.vcxproj" (default target) (1) ->
"C:\optimae\pytorch\torch\lib\build\ATen\ALL_BUILD.vcxproj" (default target) (3) ->
"C:\optimae\pytorch\torch\lib\build\ATen\src\ATen\ATen.vcxproj" (default target) (4) ->
(ClCompile target) ->
  C:\optimae\pytorch\aten\src\TH\vector\AVX2.c(60): error C2440: 'function': cannot convert from
'int' to '__m256' [C:\optimae\pytorch\torch\lib\build\ATen\src\ATen\ATen.vcxproj]

    8925 Warning(s)
    1 Error(s)
```
